crates/sui-graphql-client/src/lib.rs

```diff
@@ -0,0 +1,658 @@
+const DEVNET_HOST: &str = "https://sui-devnet.mystenlabs.com/graphql";
+const DEFAULT_LOCAL_HOST: &str = "http://localhost:9125/graphql";
+static USER_AGENT: &str = concat!(env!("CARGO_PKG_NAME"), "/", env!("CARGO_PKG_VERSION"),);
+
+#[derive(Debug)]
+/// A page of items returned by the GraphQL server.
+pub struct Page<T> {
+    /// Information about the page, such as the cursor and whether there are more pages.
+    page_info: PageInfo,
+    /// The data returned by the server.
+    data: Vec<T>,
+}
+
+impl<T> Page<T> {
+    pub fn page_info(&self) -> &PageInfo {
+        &self.page_info
+    }
+
+    pub fn data(&self) -> &[T] {
+        &self.data
+    }
+
+    fn new(page_info: PageInfo, data: Vec<T>) -> Self {
+        Self { page_info, data }
+    }
+}
+
+/// The GraphQL client for interacting with the Sui blockchain.
+/// By default, it uses the `reqwest` crate as the HTTP client.
+pub struct Client {
+    /// The URL of the GraphQL server.
+    rpc: Url,
+    /// GraphQL supports multiple versions of the service. By default, the client uses the stable
+    /// version of the service.
+    rpc_version: Option<String>,
+    inner: reqwest::Client,
+}
+
+impl Client {
+    // ===========================================================================
+    // Client Misc API
+    // ===========================================================================
+
+    /// Create a new GraphQL client with the provided server address.
+    pub fn new(server: &str) -> Result<Self, Error> {
+        let rpc = reqwest::Url::parse(server).map_err(|_| anyhow!("Invalid URL: {}", server))?;
+
+        let client = Client {
+            rpc,
+            rpc_version: None,
+            inner: reqwest::Client::builder().user_agent(USER_AGENT).build()?,
+        };
+        Ok(client)
+    }
+
+    /// Create a new GraphQL client connected to the `mainnet` GraphQL server: {MAINNET_HOST}.
+    pub fn new_mainnet() -> Self {
+        Self::new(MAINNET_HOST).expect("Invalid mainnet URL")
+    }
+
+    /// Create a new GraphQL client connected to the `testnet` GraphQL server: {TESTNET_HOST}.
+    pub fn new_testnet() -> Self {
+        Self::new(TESTNET_HOST).expect("Invalid testnet URL")
+    }
+
+    /// Create a new GraphQL client connected to the `devnet` GraphQL server: {DEVNET_HOST}.
+    pub fn new_devnet() -> Self {
+        Self::new(DEVNET_HOST).expect("Invalid devnet URL")
+    }
+
+    /// Create a new GraphQL client connected to the `localhost` GraphQL server:
+    /// {DEFAULT_LOCAL_HOST}.
+    pub fn new_localhost(&mut self) -> Self {
+        Self::new(DEFAULT_LOCAL_HOST).expect("Invalid localhost URL")
+    }
+
+    /// Set the version for the GraphQL GraphQL client. The default set version is stable.
+    ///
+    /// Use the `service_config` API to get the `available_versions` field to know which versions
+    /// are supported by this service.
+    pub fn set_version(&mut self, version: Option<&str>) -> Result<&mut Self, Error> {
```

Comment:
I'm a bit confused here and in `set_rpc_server` why this takes and returns `&mut self`/`&mut Self`. I'd expect either for this to be `&mut self`/`()` or `self`/`Self`. But maybe I'm missing something?

Reply:
Yeah the function signature is a little odd, but given the graphql version changes we can actually remove all of this!